autopatches/autopatch.py

```diff
@@ -0,0 +1,188 @@
+        git(f"remote add fork https://{login}:{token}@github.com/{login}/{reponame}", in_folder=folder)
+
+
+def apply(patch):
+
+    patch_path = os.path.abspath(os.path.join("patches", patch, "patch.sh"))
+
+    for app in progressbar(apps(), "Apply to: ", 40):
+        folder = os.path.join(".apps_cache", app["id"])
+        current_branch = git(f"symbolic-ref --short HEAD", in_folder=folder)
+        git(f"reset --hard origin/{current_branch}", in_folder=folder)
+        os.system(f"cd {folder} && bash {patch_path}")
+
+
+def diff():
+
+    for app in apps():
+        folder = os.path.join(".apps_cache", app["id"])
+        if bool(subprocess.check_output(f"cd {folder} && git diff", shell=True).strip().decode("utf-8")):
+            print("\n\n\n")
+            print("=================================")
+            print("Changes in : " + app["id"])
+            print("=================================")
+            print("\n")
+            os.system(f"cd {folder} && git --no-pager diff")
+
+
+def push(patch):
+
+    title = "[autopatch] " + open(os.path.join("patches", patch, "pr_title.md")).read().strip()
+
+    def diff_not_empty(app):
+        folder = os.path.join(".apps_cache", app["id"])
+        return bool(subprocess.check_output(f"cd {folder} && git diff", shell=True).strip().decode("utf-8"))
+
+    def app_is_on_github(app):
+        return "github.com" in app["url"]
+
+    apps_to_push = [app for app in apps() if diff_not_empty(app) and app_is_on_github(app)]
+
+    with requests.Session() as s:
+        s.headers.update({"Authorization": f"token {token}"})
+        for app in progressbar(apps_to_push, "Forking: ", 40):
+            app["repo"] = app["url"][len("https://github.com/"):].strip("/")
+            fork_if_needed(app["repo"], s)
+
+        for app in progressbar(apps_to_push, "Pushing: ", 40):
+            app["repo"] = app["url"][len("https://github.com/"):].strip("/")
+            app_repo_name = app["url"].rsplit("/", 1)[-1]
+            folder = os.path.join(".apps_cache", app["id"])
+            current_branch = git(f"symbolic-ref --short HEAD", in_folder=folder)
+            git(f"reset origin/{current_branch}", in_folder=folder)
+            git(["commit", "-a", "-m", title, "--author='Yunohost-Bot <>'"], in_folder=folder)
+            try:
+                git(f"remote remove fork", in_folder=folder)
+            except Exception:
+                pass
+            git(f"remote add fork https://{login}:{token}@github.com/{login}/{app_repo_name}", in_folder=folder)
+            git(f"push fork {current_branch}:{patch} --quiet --force", in_folder=folder)
+            create_pull_request(app["repo"], patch, current_branch, s)
+
+
+def fork_if_needed(repo, s):
+
+    repo_name = repo.split("/")[-1]
+    r = s.get(github_api + f"/repos/{login}/{repo_name}")
+
+    if r.status_code == 200:
+        return
+
+    r = s.post(github_api + f"/repos/{repo}/forks")
```

Comment:
Fork may takes time, is it ok to create it and not check if the fork is done? (I guess yes as we don't do a lot of fork compare to the amount of apps, but I prefer to ask)

Reply:
Uuuuuuh yeah good question ? I don't really know, this part of the code has not really been tested (or maybe back in the days when i first wrote this). Naively I would think the server returns 200 only after effectively creating the fork ? But maybe not ...